### ISSUE TITLE
Add location for unexpected EOF errors

### ISSUE THREAD

**Issue by brendanzab** [lalrpop/lalrpop#322]
It is sometimes helpful to be able to have the location ofthe EOF for error reporting purposes. Currently the `UnrecognizedToken` looks like this:

```rust
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum ParseError<L,T,E> {
    ...
    UnrecognizedToken {
        token: Option<(L, T, L)>,
        expected: Vec<String>
    },
    ...
}
```

What would you thoughts be about switching to this?

```rust
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum ParseError<L,T,E> {
    ...
    UnrecognizedEof {
        location: L,
        expected: Vec<String>
    },
    UnrecognizedToken {
        token: (L, T, L),
        expected: Vec<String>
    },
    ...
}
```

It's a breaking change, sadly, but I always found the old usage of `Option::None` for the EOF took a bit to wrap my head around, and that's reflected in the awkwardness of the documentation. At least here the point is more obvious in the name of the variant.

**Comment by nikomatsakis** [comment]
As I [noted on gitter](https://gitter.im/lalrpop/Lobby?at=5a90171b012ff2bf681ccba1), I am not happy with our error handling strategy. I would like to change to something based on the `Failure` care. Perhaps @withoutboats can weigh in on what they view as the Best Practices. I was thinking of first creating independent structs for all the possible LALRPOP error cases today. Then there are two possibilities of where to go from there:

- Create an enum `LalrpopError` that includes those cases, but is marked as `#[non_exhautive]` (that's not currently possible, so we can do the "unmentionable variant" trick). This enum can include a `User` variant that carries no data. We can use causal chaining to attach the user's error.
- Forego the enum altogether and just return `failure::Error`
    - This is probably not a good idea. I don't think it's best practices.

This would allow us to remove the `type Error` business altogether, I think.

**Comment by nikomatsakis** [comment]
*That said*, I guess you were talking about something more narrow. I'd be fine with having `UnrecognizedEOF` as a variant in any case. The location I guess would be the end of the final token?

**Comment by brendanzab** [comment]
Yup, that's correct!

**Comment by nikomatsakis** [comment]
@brendanzab think you would be up for opening a PR?

**Comment by brendanzab** [comment]
Might be able to look at it at some stage, perhaps tomorrow? Kind of on a role with some type system stuff though. I'll keep you posted though.

**Comment by nwtnni** [comment]
@Marwes I think we can close this issue now?

**Comment by Marwes** [comment]
Yep, I forgot to add (`Closed #322` when merging).
